### DISCUSSION TITLE
How to manage integration test, modules and autoconfiguration?

### DISCUSSION THREAD

**Discussion by michaldo** [spring-projects/spring-modulith#127] [Q&A]
Hi,

the biggest benefit of Modulith for me is opportunity to test couple of beans defined in a package.  However, I have no clear idea how to manage autoconfiguration.
Consider case I have package `order` with 3 `@Component` from order family. I can have
```
package order;
@ApplicationModuleTest
class OrderTest {
....
```
and test whole module.

Unfortunately, the test starts all autoconfigurations, including those not used by Order module. I would like to have OrderTest as fast as possible and start only required beans.
How to arrange autoconfigurations?

**Comment by genuss** [comment]
By default `@ApplicationModuleTest` starts only beans defined in the package in which the test class is defined. This is controlled by annotation [mode](https://github.com/spring-projects/spring-modulith/blob/main/spring-modulith-test/src/main/java/org/springframework/modulith/test/ApplicationModuleTest.java#L64).
So it looks like you may be experiencing a bug or some misconfiguration. Could you share an example?

**Comment by michaldo** [comment]
I had in mind case when whole application use `KafkaAutoConfiguration` or `MongoAutoConfiguration`, both triggered by class presence.
It means even if I want test only module _order_, without relation to Kafka or Mongo, I need setup (or disable) both K&M

**Comment by genuss** [comment]
I see, these configurations will be activated as soon they are somewhat "global" for the whole application. I think profiles with deactivated configuration are our friends here

**Comment by odrotbohm** [comment]
I don't think we have a good answer to that yet, as we can only control the component scanning and the auto-configuration packages for the application. We cannot (don't want to?) alter the classpath of the application, as there's no way to reasonably reason what needs to be included to run a single module. Do you suggest that there is a way in your case?
